dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIDeferredMetaDataFactoryImpl.java

```diff
@@ -0,0 +1,72 @@
+/*******************************************************************************
+ * Copyright (c) 2018 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.cdi.liberty;
+
+import java.util.Map;
+import java.util.HashMap;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+
+import com.ibm.ws.container.service.metadata.extended.DeferredMetaDataFactory;
+import com.ibm.ws.runtime.metadata.ComponentMetaData;
+
+@Component(service = {DeferredMetaDataFactory.class},
+           property = { "deferredMetaData=CDI" })
+public class CDIDeferredMetaDataFactoryImpl implements DeferredMetaDataFactory {
+
+    //Entires are removed in the finally block of the only method that sets it so it is safe from memory leaks. 
+    private static Map<String,ComponentMetaData> metaDataMap = new HashMap<String,ComponentMetaData>();
```

Comment:
First, the usage of HashMap here is not thread safe.  This structure is clearly accessed by multiple threads based on how I see you doing the puts and removes in subsequent code.
Second, it seems wrong that these entries are kept around only for such a short time within the map, when the point of implementing this is to supply context based on this information to an executor.  An executor can start tasks at any time, including the possibility of delaying to a later point.  For this approach to be valid, you would need some guarantee that the use of the executor will have at least submitted and waited for all of its tasks to have started before you can validly remove the entry from your map.  Otherwise, it is possible the entry will be needed and found to be absent.